### PR TITLE
Fix CUDA-graph capture race with DataLoader pin-memory threads

### PR DESCRIPTION
**Pull request by EHxuban11**
﻿## Summary

Fixes the CUDA-graph capture race that killed 2 of 100 datasets on the first RF100-VL Vast campaign ("AcceleratorError ... in pin memory thread").

Graph capture ran in the default `capture_error_mode="global"`, under which a CUDA call from **any** host thread invalidates the capture. A DataLoader with `pin_memory=True` keeps a pin-memory thread alive that calls `cudaHostAlloc` whenever it stages the next prefetched batch, so a lazy capture taken mid-epoch races with it: the capture dies with `cudaErrorStreamCaptureUnsupported` and, worse, the error poisons the pin-memory thread, which re-raises on the next batch fetch and kills the whole run.

`thread_local` is PyTorch's documented mode for exactly this situation: only the capturing thread is restricted during capture; other threads' CUDA calls proceed normally without being captured. The pin-memory thread touches only its own pinned host buffers, never the capturing stream, so nothing it does belongs in the graph.

- **Inference/validation (`GraphRunner._capture`)**: pass `capture_error_mode="thread_local"` to `torch.cuda.graph` directly.
- **Training (`TrainGraphManager._capture_and_run`)**: `torch.cuda.make_graphed_callables` constructs its own `CUDAGraph` objects and exposes no way to choose the mode, so the `torch.cuda.CUDAGraph` symbol is swapped for a mode-forcing subclass for the duration of the call (restored in a `finally`).

No behavior change otherwise: capture points, fallback ladders, and numerics are untouched (the existing eager/graphed parity suites still gate bit-identical loss trajectories).

## Test plan

- [x] New CPU tests: the patch context swaps/restores `torch.cuda.CUDAGraph` (including on exception), forces `thread_local` at `capture_begin`, and the manager's capture path instantiates the patched class.
- [x] New GPU regression tests on both paths: capture succeeds while a stand-in pin-memory thread hammers real `cudaHostAlloc` (strictly growing sizes to defeat the caching host allocator), and the thread is never poisoned.
- [x] Negative control on a local RTX GPU: the same hammer against `capture_error_mode="global"` reproduces the exact campaign failure ÔÇö hammer thread dies with `AcceleratorError` / `cudaErrorStreamCaptureUnsupported` and the capture is invalidated.
- [x] `test_cuda_graph_training.py` + `test_cuda_graph.py`: 68 passed, 1 skipped (needs 2 GPUs), 1 deselected (external_data).
- [x] `test_validator_cuda_graph.py`, `test_validator_reuse.py`, `test_image_cache.py`: 25 passed.

## Code provenance

All code in this PR is original, written for this PR. It uses only PyTorch's public, documented API (`capture_error_mode="thread_local"` on `torch.cuda.graph` / `CUDAGraph.capture_begin`); no third-party code was ported, adapted, or derived.

<details><summary><h3>Greptile Summary</h3></summary>

The PR prevents DataLoader pin-memory activity from invalidating CUDA graph capture and completes the earlier concurrency fix by serializing the temporary training-time CUDAGraph replacement.
- Uses thread-local capture error handling for inference and validation graphs.
- Temporarily substitutes a mode-forcing CUDAGraph subclass during training capture, with locking and unconditional restoration.
- Adds CPU concurrency/restoration tests and GPU regression coverage for concurrent pinned-memory allocation.
</details>

<details><summary><h3>Confidence Score: 5/5</h3></summary>

The PR appears safe to merge.

No blocking failure remains; the previously reported unsynchronized global class replacement is now serialized and restored in a finally block.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| libreyolo/training/cuda_graph.py | Adds a serialized, exception-safe temporary CUDAGraph subclass that forces thread-local capture errors during training graph construction. |
| libreyolo/models/base/cuda_graph.py | Configures inference and validation graph capture to use thread-local error handling directly. |
| tests/unit/test_cuda_graph_training.py | Covers symbol replacement, restoration, concurrent serialization, forced capture mode, manager integration, and pinned-allocation regression behavior. |
| tests/unit/test_cuda_graph.py | Adds GPU regression coverage proving inference capture survives concurrent pinned-memory allocations. |

</details>

<details open><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[Training graph capture requested] --> B[Acquire capture patch lock]
    B --> C[Save original CUDAGraph class]
    C --> D[Install thread-local forcing subclass]
    D --> E[make_graphed_callables captures graph]
    E --> F[Restore original class in finally]
    F --> G[Release capture patch lock]
    H[DataLoader pin-memory thread] --> I[cudaHostAlloc proceeds independently]
    I -. does not invalidate .-> E
```
</details>

<sub>Reviews (2): Last reviewed commit: ["Unbreak dev CI: moge2 export support is ..."](https://github.com/libreyolo/libreyolo/commit/59ce05137aa003376f6b92321dbb4f89239af737) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=49387033)</sub>

### CODE REVIEW

File: libreyolo/training/cuda_graph.py [hunk 1]
```diff
@@ -55,14 +55,54 @@
...
+    class _ThreadLocalCaptureGraph(original):  # type: ignore[misc, valid-type]
+        def capture_begin(self, *args: Any, **kwargs: Any) -> None:
+            kwargs["capture_error_mode"] = "thread_local"
+            super().capture_begin(*args, **kwargs)
+
+    torch.cuda.CUDAGraph = _ThreadLocalCaptureGraph
+    try:
+        yield
+    finally:
+        torch.cuda.CUDAGraph = original
```
[the remark this answers is not in the document]
Comment (EHxuban11) [reply libreyolo/training/cuda_graph.py]: Valid, fixed in 27986b8f.

A module-level `threading.Lock` now serializes the whole context, so overlapping captures can no longer interleave their save/restore of `torch.cuda.CUDAGraph` (previously the second thread saved the first thread's subclass as its "original", and depending on exit order a temporary subclass could stay installed for the rest of the process). Holding the lock across the capture is fine: captures are rare, and two truly concurrent same-process captures were never a supported pattern anyway.

Added a regression test that runs four overlapping contexts and asserts each one sees the true original as its base class and that the true original is restored at the end. Verified it fails without the lock and passes with it.
